[JSC][armv7] Reserve `d14` as a scratch register for wasm JITs

jjgriego opened the pull request:
#### 1ff6d45f35c140efefa6110b5d5a71a872f3821e
<pre>
[JSC][armv7] Reserve `d14` as a scratch register for wasm JITs
<a href="https://bugs.webkit.org/show_bug.cgi?id=263324">https://bugs.webkit.org/show_bug.cgi?id=263324</a>

Reviewed by Justin Michaud.

We really need an extra FP scratch register for wasm BBQ on armv7. In
particular, we use the `nonPreservedNonArgumentFPR0` for this purpose, but we
also ended up using register currently singled out as the
`nonPreservedNonArgumentFPR0`, `d7` as an argument register in BBQ JIT contexts.

To avoid the overlap, reserve `d14` as a scratch register--we copy exactly the
approach used to reserve `d15` as a scratch FPR for the macroassembler [1]:
though the register is callee-save in the platform ABI, we save/restore only at
VM entry/exit and the register is volatile in VM contexts.

[1] <a href="https://commits.webkit.org/250952@main">https://commits.webkit.org/250952@main</a>

* Source/JavaScriptCore/jit/FPRInfo.h:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::vmCalleeSaveRegisters):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:

Canonical link: <a href="https://commits.webkit.org/269762@main">https://commits.webkit.org/269762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a864f296d9fe6c9032815ea90aa5a901f2bdc316

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23766 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24588 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20998 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1432 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23209 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21938 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/220 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19682 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25441 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20553 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26792 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19753 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20593 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20796 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24630 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22045 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/230 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18085 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/26097 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/179 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6119 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5610 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/307 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27379 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/240 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5945 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
macroClobberedFPRs etc. needs to be changed. Can you grep all `fpTempRegister` sites, and fix that? I think we should define `fpTempRegister2` for ARMv7 and use it in the same way to RISCV.

justinmichaud reviewed (approved):
r=me

jjgriego committed:
[JSC][armv7] Reserve `d14` as a scratch register for wasm JITs
https://bugs.webkit.org/show_bug.cgi?id=263324

Reviewed by Justin Michaud.

We really need an extra FP scratch register for wasm BBQ on armv7. In
particular, we use the `nonPreservedNonArgumentFPR0` for this purpose, but we
also ended up using register currently singled out as the
`nonPreservedNonArgumentFPR0`, `d7` as an argument register in BBQ JIT contexts.

To avoid the overlap, reserve `d14` as a scratch register--we copy exactly the
approach used to reserve `d15` as a scratch FPR for the macroassembler [1]:
though the register is callee-save in the platform ABI, we save/restore only at
VM entry/exit and the register is volatile in VM contexts.

[1] https://commits.webkit.org/250952@main

* Source/JavaScriptCore/jit/FPRInfo.h:
* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::vmCalleeSaveRegisters):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:

Canonical link: https://commits.webkit.org/269762@main
